### PR TITLE
make inbox unboxing interruptible, and more aggressively limit parallelism CORE-8968

### PR DESCRIPTION
**Pull request by mmaxim**
Patch does the following:

1.) Refactor "localizing" inbox items into a its own file.
2.) Make `InboxSource` have a single `localizePipeline` per instance.
3.) Make `localizePipeline` interruptible.
4.) Have `GetThreadNonblock` interrupt it so that when we fetch a thread there is no contention for resources.

### CODE REVIEW

File: go/chat/localizer.go [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+	}
+	return res
+}
+
+func (b *baseLocalizer) getConvs(inbox types.Inbox, maxLocalize *int) []chat1.Conversation {
+	convs := utils.PluckConvs(inbox.ConvsUnverified)
+	if maxLocalize == nil {
+		return convs
+	}
+	return convs[:*maxLocalize]
```
Comment (joshblum) [inline go/chat/localizer.go]: should check the length here or could out of bounds panic

File: go/chat/localizer.go [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+	maxLocalize *int) (res []chat1.ConversationLocal, err error) {
+	defer b.Trace(ctx, func() error { return err }, "Localize")()
+	inbox = b.filterSelfFinalized(ctx, inbox)
+	convs := b.getConvs(inbox, maxLocalize)
+	b.baseLocalizer.pipeline.queue(ctx, uid, convs, b.localizeCb)
+	if err != nil {
+		return res, err
+	}
+	res = make([]chat1.ConversationLocal, len(convs))
+	indexMap := make(map[string]int)
```
Comment (joshblum) [inline go/chat/localizer.go]: why does the ordering matter here?

File: go/chat/localizer.go [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+		localizeCb:    localizeCb,
+	}
+}
+
+func (b *blockingLocalizer) Localize(ctx context.Context, uid gregor1.UID, inbox types.Inbox,
+	maxLocalize *int) (res []chat1.ConversationLocal, err error) {
+	defer b.Trace(ctx, func() error { return err }, "Localize")()
+	inbox = b.filterSelfFinalized(ctx, inbox)
+	convs := b.getConvs(inbox, maxLocalize)
+	b.baseLocalizer.pipeline.queue(ctx, uid, convs, b.localizeCb)
```
Comment (joshblum) [inline go/chat/localizer.go]: should this be returning an error on the queue being full? we check `err` below but it's from the return value

File: go/chat/localizer.go [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+	filteredInbox := b.filterInboxRes(ctx, inbox, uid)
+	// Send inbox over localize channel
+	b.localizeCb <- types.AsyncInboxResult{
+		InboxRes: &filteredInbox,
+	}
+	// Spawn off localization into its own goroutine and use cb to communicate with outside world
+	bctx := BackgroundContext(ctx, b.G())
+	go func() {
+		b.Debug(bctx, "Localize: starting background localization: convs: %d", len(inbox.ConvsUnverified))
+		b.baseLocalizer.pipeline.queue(bctx, uid, b.getConvs(inbox, maxLocalize), b.localizeCb)
```
Comment (joshblum) [inline go/chat/localizer.go]: check and log error? (if this returns an error as noted above)

File: go/chat/localizer.go [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+	retCh chan types.AsyncInboxResult) {
+	defer s.Trace(ctx, func() error { return nil }, "queue")()
+	s.Lock()
+	defer s.Unlock()
+	job := newLocalizerPipelineJob(ctx, s.G(), uid, convs, retCh)
+	job.ctx, job.cancelFn = context.WithCancel(BackgroundContext(ctx, s.G()))
+	select {
+	case s.jobQueue <- job:
+	default:
+		s.Debug(ctx, "queue: dropped job, queue full")
```
Comment (joshblum) [inline go/chat/localizer.go]: return this as an err?

File: go/chat/localizer.go [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+func (s *localizerPipeline) resume(ctx context.Context) bool {
+	defer s.Trace(ctx, func() error { return nil }, "resume")()
+	s.Lock()
+	defer s.Unlock()
+	s.suspendWg.Done()
+	return false
+}
+
+func (s *localizerPipeline) isContextCanceled(err error) bool {
+	if err == nil {
```
Comment (joshblum) [inline go/chat/localizer.go]: can we drop this check and let it fall through to the end?

File: go/chat/localizer.go [hunk 7]
```diff
@@ -0,0 +1,952 @@
...
+					s.Debug(job.ctx, "localizeLoop: re-enqueuing canceled job")
+					s.jobQueue <- job.retry(s.G())
+				}
+				if job.closeIfDone() {
+					s.Debug(job.ctx, "localizeLoop: all job tasks complete")
+				}
+			}()
+			select {
+			case <-doneCh:
+				job.cancelFn()
```
Comment (joshblum) [inline go/chat/localizer.go]: why do we call this here?

File: go/chat/localizer.go [hunk 8]
```diff
@@ -0,0 +1,952 @@
...
+
+	return foundMsgs, chat1.ConversationErrorType_NONE, nil
+}
+
+func (s *localizerPipeline) getMinWriterRoleInfoLocal(ctx context.Context, info *chat1.ConversationMinWriterRoleInfo) *chat1.ConversationMinWriterRoleInfoLocal {
+	if info == nil {
+		return nil
+	}
+	username := ""
+	name, err := s.G().GetUPAKLoader().LookupUsername(ctx, keybase1.UID(info.Uid.String()))
```
Comment (joshblum) [inline go/chat/localizer.go]: realize this is just a mv but should we use uid mapper here instead for performance?

File: go/chat/localizer.go [hunk 9]
```diff
@@ -0,0 +1,952 @@
...
+	select {
+	case s.jobQueue <- job:
+	default:
+		s.Debug(ctx, "queue: dropped job, queue full")
+	}
+	s.Debug(ctx, "queue: len: %d", len(s.jobQueue))
+}
+
+func (s *localizerPipeline) clearQueue() {
+	s.jobQueue = make(chan *localizerPipelineJob, 500)
```
Comment (joshblum) [inline go/chat/localizer.go]: can you make this configurable and drop it to 1-2 in tests to see if any bad behavior is triggered?

File: go/chat/localizer.go [hunk 10]
```diff
@@ -0,0 +1,952 @@
...
+		conv  chat1.Conversation
+		index int
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	convCh := make(chan job)
+	pending := localizeJob.getPending()
+	retCh := make(chan chat1.ConversationID, len(pending))
+
+	if len(pending) == 0 {
```
Comment (joshblum) [inline go/chat/localizer.go]: check before making `retCh`

File: go/chat/localizer.go (hunk 9 above)
Comment (mmaxim) [reply go/chat/localizer.go]: I'll just make `queue` block if the buffer is full, think it is ok.

**Comment by mmaxim** [comment]
ok @joshblum I think that addresses your current comments.

**Pull request merged**